packages/flutter/lib/src/widgets/_web_image_web.dart

```diff
@@ -350,19 +350,90 @@ class RenderWebImage extends RenderShiftedBox {
+      // boxes. The offset of the child needs to be worked out from the
+      // alignment. In the `paint` method, a clip will be applied so that only
+      // the portion of the child that is in the destination box will be shown.
+      final Size sourceSize = fittedSizes.source;
+      final Size destinationSize = fittedSizes.destination;
+
+      // Only `BoxFit.fill` distorts the aspect ratio of the image. The rest of
+      // the `BoxFit` values preserve the aspect ratio. Furthermore, if the
+      // `fit` is `BoxFit.fill` then `fittedSizes.source == inputSize` since the
+      // entire image is drawn into the box. So the case of `BoxFit.fill` is
+      // never hit in this branch of the conditional. Therefore, the child
+      // size should be the source size multiplied by a scalar value. The scale
+      // can be determined by examining the relative sizes of the `sourceSize`
+      // and `destinationSize`. Since aspect ratio is preserved, `sourceSize`
+      // and `destinationSize` must have the same aspect ratio. Therefore, the
+      // scale to multiply the `inputSize` to get the `childSize` is equal to
+      // the ratio of the `destinationSize` to the `sourceSize`.
+      assert(sourceSize.aspectRatio == destinationSize.aspectRatio);
+      final double scale = destinationSize.width / sourceSize.width;
+      final Size childSize = inputSize * scale;
+      child!.layout(BoxConstraints.tight(childSize));
+
+      // Now that the child has been sized, it must also be offset by an
+      // appropriate amount so that the specified portion of the image is shown
+      // in the destination box. `Alignment.inscribe` gives the sub-rectangle
+      // of source size that will be shown in the destination box. Again, the
+      // source and destination box have the same aspect ratio, so to compute
+      // the offset for the child, first compute the offset that the original
+      // sized image would need to put the sub-rectangle given by
+      // `Alignment.inscribe` at the origin, and then scale that offset by the
+      // same scale computed above to get the offset for the child.
+      final Rect sourceRect = _resolvedAlignment!.inscribe(sourceSize, Offset.zero & inputSize);
+      final Offset childOffset = Offset(-sourceRect.left, -sourceRect.top) * scale;
+      final BoxParentData childParentData = child!.parentData! as BoxParentData;
+      childParentData.offset = childOffset;
+    }
+  }
+
+  @override
+  void paint(PaintingContext context, Offset offset) {
+    if (child == null) {
+      return;
+    }
+    final Size inputSize = Size(image.naturalWidth.toDouble(), image.naturalHeight.toDouble());
+    fit ??= BoxFit.scaleDown;
+    final FittedSizes fittedSizes = applyBoxFit(fit!, inputSize, size);
+    final Rect destinationRect = offset & fittedSizes.destination;
+    context.pushClipRect(needsCompositing, offset, destinationRect, super.paint);
```

Comment:
I feel like we are applying the `offset` twice, aren't we?

`pushClipRect` also shifts the rect by the given offset:
https://github.com/flutter/flutter/blob/43657f3baa17bc1df6cc15da8b0fea2290c2c4fd/packages/flutter/lib/src/rendering/object.dart#L566